File: es15/es15.py
#!/usr/bin/env python3

import sys
import copy

class Unit(object):
    def __init__(self, cls, x, y):
        self.cls = cls
        self.x = x
        self.y = y
        self.hp = 200
        self.attack = 3

    def __lt__(self, other):
        if self.y != other.y:
            return self.y < other.y

        return self.x < other.x

class Map(object):
    CLSDSC = {"E": "Elves", "G": "Goblins"}

    def __init__(self):
        self.width = self.height = 0
        self.units = []
        self.data = None
        self.dist = None

    def load(self, f):
        self.data = []
        self.units = []
        y = 0
        for line in f:
            row = []
            x = 0
            for b in line:
                if b == 'E' or b == 'G':
                    self.units.append(Unit(b, x, y))
                    b = '.'
                if b == '.' or b == '#':
                    row.append(b)
                    x += 1

            self.data.append(row)
            y += 1

        self.height = len(self.data)
        self.width = len(self.data[0])

    def adjacent(self, x, y):
        return ((x, y-1), (x-1, y), (x+1, y), (x, y+1))

    def place_units(self):
        for u in self.units:
            if u.hp > 0:
                self.data[u.y][u.x] = u.cls

    def remove_units(self):
        for u in self.units:
            if u.hp > 0:
                self.data[u.y][u.x] = '.'

    def bfs(self, x, y):
        self.place_units()

        maxdist = self.height * self.width
        self.dist = [[maxdist for x in range(self.width)] for y in range(self.height)]

        queue = [(x, y)]
        self.dist[y][x] = 0
        while len(queue) > 0:
            x, y = queue.pop(0)
            for x1, y1 in self.adjacent(x, y):
                if self.data[y1][x1] != '.':
                    continue
                if self.dist[y1][x1] == maxdist:
                    self.dist[y1][x1] = self.dist[y][x] + 1
                    queue.append((x1, y1))

        self.remove_units()

    def find_target(self, unit):
        self.bfs(unit.x, unit.y)
        mindist = self.height * self.width
        target = None
        for e in self.units:
            if e.cls != unit.cls and e.hp > 0:
                for x, y in self.adjacent(e.x, e.y):
                    if mindist > self.dist[y][x]:
                        mindist = self.dist[y][x]
                        target = (x, y)

        return target

    def move(self, unit):
        stop = True
        for e in self.units:
            if e.hp > 0 and e.cls != unit.cls:
                stop = False
                break

        if stop:
            return False

        target = self.find_target(unit)
        if target is None:
            return True

        tx, ty = target
        if tx == unit.x and ty == unit.y:
            return True

        while True:
            mindist = self.width * self.height
            minx, miny = tx, ty
            for x, y in self.adjacent(tx, ty):
                if mindist > self.dist[y][x]:
                    mindist = self.dist[y][x]
                    minx, miny = x, y

            if minx == unit.x and miny == unit.y:
                break

            tx, ty = minx, miny

        unit.x = tx
        unit.y = ty
        return True

    def attack(self, unit):
        minhp = 201
        target = None
        adj = self.adjacent(unit.x, unit.y)
        for e in self.units:
            if e.cls != unit.cls and e.hp > 0:
                for x, y in adj:
                    if x == e.x and y == e.y and minhp > e.hp:
                        minhp = e.hp
                        target = e
                        break

        if target:
            target.hp -= unit.attack

    def playround(self):
        self.units.sort()
        for u in self.units:
            if u.hp > 0:
                if not self.move(u):
                    return False

                self.attack(u)

        return True

    def get_survivors_cls(self):
        for u in self.units:
            if u.hp > 0:
                return self.CLSDSC[u.cls]
        return ""

    def get_hp(self):
        return sum(x.hp for x in self.units if x.hp > 0)

    def playgame(self, retearly = False):
        i = 0
        while self.playround():
            if retearly:
                for u in self.units:
                    if u.hp < 0 and u.cls == "E":
                        return self.CLSDSC["G"], i, self.get_hp()
            i += 1

        return self.get_survivors_cls(), i, self.get_hp()

    def __str__(self):
        self.place_units()
        units = sorted(self.units)
        rows = []
        for y in range(self.height):
            data = ["".join(self.data[y])]

            for u in units:
                if u.y == y and u.hp > 0:
                    data.append("  {}({})".format(u.cls, u.hp))

            rows.append("".join(data))

        self.remove_units()
        return "\n".join(rows)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file = sys.stderr)
        exit(-1)

    m = Map()
    try:
        with open(sys.argv[1], "rt") as f:
            m.load(f)
    except:
        raise
        print("Cannot open {} for reading".format(sys.argv[1]), file = sys.stderr)
        exit(-1)

    m1 = copy.deepcopy(m)
    cls, rounds, hp = m1.playgame()
    print(m1)
    print("Combat ends after {} full rounds".format(rounds))
    print("{} win with {} total hit points left".format(cls, hp))
    print("Outcome: {} * {} = {}".format(rounds, hp, rounds * hp))

    print("\nBisecting the attack level for Elves")
    low = 4
    high = 200
    lastm, lastrounds, lasthp = None, None, None
    while low < high:
        mid = (low + high) // 2
        print("Trying with attack {}...".format(mid))
        m1 = copy.deepcopy(m)
        for u in m1.units:
            if u.cls == 'E':
                u.attack = mid

        cls, rounds, hp = m1.playgame(True)
        if cls == m1.CLSDSC["G"]:
            low = mid + 1
        else:
            high = mid

            lastm = m1
            lastrounds = rounds
            lasthp = hp

    if lastm:
        print(lastm)
        print("Minumum attack needed without dead Elves is {}".format(high))
        print("Combat ends after {} full rounds".format(lastrounds))
        print("Elves win with {} total hit points left".format(lasthp))
        print("Outcome: {} * {} = {}".format(lastrounds, lasthp, lastrounds * lasthp))

File: es15/es15.c
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum {
	CLASS_ELF,
	CLASS_GOBLIN,
};

const int off[][2] = {
	/* X,  Y */
	{  0, -1 },
	{ -1,  0 },
	{  1,  0 },
	{  0,  1 },
};

struct unit
{
	int class;
	int hitpoints;
	int attack;
	size_t x, y;
};

struct playground
{
	char **data;
	size_t height;
	size_t width;
	size_t dsize;

	struct unit *units;
	size_t ucount;
	size_t usize;

	int **dist;
};

static void playground_free(struct playground *g)
{
	if (g) {
		free(g->units);
		while (g->height-- > 0) {
			free(g->dist[g->height]);
			free(g->data[g->height]);
		}
		free(g->dist);
		free(g->data);
		free(g);
	}
}

static struct playground *playground_copy(struct playground *g)
{
	struct playground *c = calloc(1, sizeof(*c));
	if (c) {
		c->dsize = c->height = g->height;
		c->width = g->width;
		c->usize = c->ucount = g->ucount;

		c->data = calloc(c->dsize, sizeof(c->data[0]));
		c->dist = calloc(c->dsize, sizeof(c->dist[0]));
		c->units = calloc(c->usize, sizeof(c->units[0]));
		if (!c->data || !c->dist || !c->units) {
			playground_free(c);
			return NULL;
		}

		memmove(c->units, g->units, c->ucount * sizeof(c->units[0]));
		for (size_t i = 0; i < c->height; i++) {
			c->data[i] = strdup(g->data[i]);
			c->dist[i] = malloc(c->width * sizeof(c->dist[0][0]));
			if (!c->data[i] || !c->dist[i]) {
				playground_free(c);
				return NULL;
			}
			memmove(c->dist[i], g->dist[i], c->width * sizeof(c->dist[0][0]));
		}
	}
	return c;
}

static void playground_bfs(struct playground *g, struct unit *u)
{
	/* reset the map */
	for (size_t y = 0; y < g->height; y++)
		for (size_t x = 0; x < g->width; x++)
			g->dist[y][x] = INT_MAX;

	/* mark the units as walls in the map */
	for (struct unit *e = g->units; e < g->units + g->ucount; e++) {
		if (e->hitpoints > 0 && u != e)
			g->data[e->y][e->x] = '#';
	}

	struct pos {
		size_t x, y;
	} queue[1000];
	size_t qcount = 0;
	queue[qcount++] = (struct pos){u->x, u->y};
	g->dist[u->y][u->x] = 0;
	while (qcount > 0) {
		struct pos p = queue[0];
		qcount--;
		memmove(queue, queue+1, qcount*sizeof(queue[0]));
		for (int i = 0; i < 4; i++) {
			size_t x = p.x + off[i][0];
			size_t y = p.y + off[i][1];

			if (g->data[y][x] != '.')
				continue;

			if (g->dist[y][x] == INT_MAX) {
				g->dist[y][x] = g->dist[p.y][p.x] + 1;
				if (qcount == 1000)
					abort();
				queue[qcount++] = (struct pos){x, y};
			}
		}
	}

	/* remove the units from the map */
	for (struct unit *e = g->units; e < g->units + g->ucount; e++) {
		if (e->hitpoints > 0 && u != e)
			g->data[e->y][e->x] = '.';
	}
}

static int playground_find_target(struct playground *g, struct unit *u, size_t *px, size_t *py)
{
	playground_bfs(g, u);
	int min = INT_MAX;
	for (struct unit *e = g->units; e < g->units + g->ucount; e++) {
		if (u->class == e->class)
			continue;

		if (e->hitpoints <= 0)
			continue;

		for (int i = 0; i < 4; i++) {
			size_t ex = e->x + off[i][0];
			size_t ey = e->y + off[i][1];
			if (min > g->dist[ey][ex]) {
				min = g->dist[ey][ex];
				*px = ex;
				*py = ey;
			}
		}
	}

	return min != INT_MAX;
}

static int playground_move_unit(struct playground *g, struct unit *u)
{
	struct unit *e;
	for (e = g->units; e < g->units + g->ucount; e++) {
		if (e->hitpoints > 0 && e->class != u->class)
			break;
	}
	if (e == g->units + g->ucount)
		return 0;

	size_t tx, ty;
	if (!playground_find_target(g, u, &tx, &ty))
		return 1;

	if (tx == u->x && ty == u->y)
		return 1;

	for (;;) {
		int min = INT_MAX;
		size_t minx = tx;
		size_t miny = ty;
		for (int i = 0; i < 4; i++) {
			size_t x = tx + off[i][0];
			size_t y = ty + off[i][1];
			if (min > g->dist[y][x]) {
				min = g->dist[y][x];
				minx = x;
				miny = y;
			}
		}
		if (minx == u->x && miny == u->y)
			break;

		tx = minx;
		ty = miny;
	}

	u->x = tx;
	u->y = ty;
	return 1;
}

static int playground_attack_unit(struct playground *g, struct unit *u)
{
	struct unit *mine = NULL;
	int minhp = INT_MAX;
	for (int i = 0; i < 4; i++) {
		for (struct unit *e = g->units; e < g->units + g->ucount; e++) {
			/* don't attack friends */
			if (e->class == u->class)
				continue;

			/* don't attack dead units */
			if (e->hitpoints <= 0)
				continue;

			if (u->x + off[i][0] == e->x &&
			    u->y + off[i][1] == e->y &&
			    minhp > e->hitpoints) {
				minhp = e->hitpoints;
				mine = e;
			}
		}
	}

	if (mine) {
		mine->hitpoints -= u->attack;
		return 1;
	}

	return 0;
}

static int unit_cmp(const void *pa, const void *pb)
{
	const struct unit *a = pa;
	const struct unit *b = pb;
	if (a->y != b->y)
		return a->y - b->y;

	return a->x - b->x;
}

static int playground_points(struct playground *g)
{
	int count = 0;
	for (struct unit *u = g->units; u < g->units + g->ucount; u++) {
		if (u->hitpoints > 0)
			count += u->hitpoints;
	}
	return count;
}

static void playground_print(struct playground *g)
{
	for (size_t y = 0; y < g->height; y++) {
		for (size_t x = 0; x < g->width; x++) {
			struct unit *u;
			for (u = g->units; u < g->units + g->ucount; u++) {
				if (u->hitpoints > 0 && x == u->x && y == u->y) {
					printf(" %c", u->class == CLASS_ELF ? 'E' : 'G');
					break;
				}
			}
			if (u == g->units + g->ucount) {
				printf(" %c", g->data[y][x]);
			}
		}

		for (struct unit *u = g->units; u < g->units + g->ucount; u++) {
			if (u->hitpoints > 0 && u->y == y) {
				printf(" %c(%d)", u->class == CLASS_ELF ? 'E' : 'G', u->hitpoints);
			}
		}
		printf("\n");
	}
}

static int playground_simulate(struct playground *c, int retearly)
{
	int rounds = 0;
	for (rounds = 0; ; rounds++) {
		/* sort the units */
		qsort(c->units, c->ucount, sizeof(c->units[0]), unit_cmp);

		/* move and attack */
		for (struct unit *u = c->units; u < c->units + c->ucount; u++) {
			if (u->hitpoints > 0) {
				if (!playground_move_unit(c, u))
					return rounds;

				playground_attack_unit(c, u);
			}
		}

		/* check if elf dead */
		if (retearly) {
			for (struct unit *u = c->units; u < c->units + c->ucount; u++)
				if  (u->class == CLASS_ELF && u->hitpoints <= 0)
					return rounds;
		}
	}
}

static struct playground *playground_load(FILE *input)
{
	struct playground *g = calloc(1, sizeof(*g));
	if (!g)
		return NULL;

	char *line = NULL;
	size_t linesize = 0;
	while (getline(&line, &linesize, input) != -1) {
		for (char *t = line; *t; t++) {
			int class;
			switch(*t) {
			case 'G': class = CLASS_GOBLIN; break;
			case 'E': class = CLASS_ELF; break;
			case '\n':
				g->width = (size_t)(t - line);
				*t = 0;
				continue;
			default:
				continue;
			}
			if (g->ucount == g->usize) {
				size_t newsize = g->usize ? g->usize * 2 : 2;
				struct unit *newunits = realloc(g->units, newsize * sizeof(*newunits));
				if (newunits == NULL) {
					playground_free(g);
					return NULL;
				}
				g->usize = newsize;
				g->units = newunits;
			}
			g->units[g->ucount].hitpoints = 200;
			g->units[g->ucount].attack = 3;
			g->units[g->ucount].class = class;
			g->units[g->ucount].x = (size_t)(t - line);
			g->units[g->ucount].y = g->height;
			g->ucount++;
			*t = '.';
		}

		if (g->height == g->dsize) {
			size_t newsize = g->dsize ? g->dsize * 2 : 2;
			char **newdata = realloc(g->data, newsize * sizeof(*newdata));
			if (newdata == NULL) {
				playground_free(g);
				return NULL;
			}
			g->dsize = newsize;
			g->data = newdata;

			int **newdist = realloc(g->dist, newsize * sizeof(*newdist));
			if (newdist == NULL) {
				playground_free(g);
				return NULL;
			}
			g->dist = newdist;
		}
		g->data[g->height] = strdup(line);
		g->dist[g->height] = malloc(g->width * sizeof(int));
		if (g->data[g->height] == NULL || g->dist[g->height] == NULL) {
			free(g->dist[g->height]);
			free(g->data[g->height]);
			playground_free(g);
			return NULL;
		}
		g->height++;
	}
	free(line);

	return g;
}

int main(int argc, char *argv[])
{
	if (argc < 2) {
		fprintf(stderr, "Usage: %s <filename>\n", argv[0]);
		return -1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input) {
		fprintf(stderr, "Cannot open %s for reading\n", argv[1]);
		return -1;
	}

	struct playground *g = playground_load(input);
	fclose(input);

	printf("Read a map of size %zu,%zu, units %zu\n", g->width, g->height, g->ucount);
	printf("Answer1:\n");
	struct playground *c = playground_copy(g);
	int rounds = playground_simulate(c, 0);
	playground_print(c);
	int points = playground_points(c);
	printf("Rounds %d, points %d\n", rounds, points);
	printf("Outcome of the battle %d with attack level = 3\n", rounds * points);
	playground_free(c);

	/* find the minimum attack level by bisecting */
	c = NULL;
	int low = 4, high = 200;
	while (low < high) {
		playground_free(c);
		c = playground_copy(g);
		int attack = (low + high) / 2;
		for (struct unit *u = c->units; u < c->units + c->ucount; u++)
			if (u->class == CLASS_ELF)
				u->attack = attack;

		printf("Trying with attack level %d...\n", attack);
		rounds = playground_simulate(c, 1);
		points = playground_points(c);
		struct unit *u;
		for (u = c->units; u < c->units + c->ucount; u++)
			if (u->class == CLASS_ELF && u->hitpoints <= 0)
				break;
		if (u < c->units + c->ucount) {
			low = attack + 1;
		} else {
			high = attack;
		}
	}
	playground_print(c);
	playground_free(c);
	for (struct unit *u = g->units; u < g->units + g->ucount; u++)
		if (u->class == CLASS_ELF)
			u->attack = high;
	rounds = playground_simulate(g, 0);
	points = playground_points(g);
	printf("Minimum attack level: %d\n", high);
	printf("Rounds %d, points %d\n", rounds, points);
	printf("Outcome of the battle %d with attack level = %d\n", rounds * points, high);

	playground_free(g);
	return 0;
}
